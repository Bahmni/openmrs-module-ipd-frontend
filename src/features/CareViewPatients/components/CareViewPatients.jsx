import React from "react";
import "../styles/CareViewPatients.scss";
import { Accordion, AccordionItem, Dropdown } from "carbon-components-react";

export const CareViewPatients = () => {
  return (
    <div className="care-view-patients-container">
      <div className="care-view-patients">
        <div className="task-type">
          <Dropdown
            id="default"
            label="Dropdown menu options"
            items={["All tasks", "Pending", "Done"]}
            itemToString={(item) => (item ? item : "")}
            selectedItem={"All tasks"}
          />
        </div>
        <Accordion className="patient-section">
          <AccordionItem title="Patient 1">Patient 1</AccordionItem>
          <AccordionItem title="Patient 2">Patient 2</AccordionItem>
        </Accordion>
      </div>
    </div>
  );
};
