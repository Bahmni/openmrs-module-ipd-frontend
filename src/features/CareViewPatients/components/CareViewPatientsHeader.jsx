import React from "react";
import PropTypes from "prop-types";
import "../styles/CareViewPatientsHeader.scss";
import { IPD_WARD_SEARCH_PLACEHOLDER_TEXT } from "../../../constants";
import { Dropdown, Search } from "carbon-components-react";

export const CareViewPatientsHeader = (props) => {
  const { handleKeyPress, handleClear, searchValue, updateSearchValue } = props;

  const handleSearchOnChange = (e) => {
    updateSearchValue(e.target.value);
  };

  return (
    <div className="task-type">
      <Search
        placeholder={IPD_WARD_SEARCH_PLACEHOLDER_TEXT}
        size="lg"
        id="ipd-ward-search"
        onChange={handleSearchOnChange}
        value={searchValue}
        onKeyDown={handleKeyPress}
        onClear={handleClear}
        className="ward-patient-search"
        labelText="Patient Search"
        light={true}
      />
      <Dropdown
        id="default"
        label="Dropdown menu options"
        items={["All tasks", "Pending", "Done"]}
        itemToString={(item) => (item ? item : "")}
        selectedItem={"All tasks"}
        light={true}
      />
    </div>
  );
};

CareViewPatientsHeader.propTypes = {
  handleKeyPress: PropTypes.func,
  handleClear: PropTypes.func,
  searchValue: PropTypes.string,
  updateSearchValue: PropTypes.func,
};
