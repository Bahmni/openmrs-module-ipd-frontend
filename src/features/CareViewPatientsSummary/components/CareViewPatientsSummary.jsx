import React, { useEffect, useState } from "react";
import propTypes from "prop-types";
import "../styles/CareViewPatientsSummary.scss";
import { HospitalBed16 } from "@carbon/icons-react";
import { Link } from "carbon-components-react";
import { FormattedMessage } from "react-intl";
import { getSlotsForPatientsAndTime } from "../../CareViewSummary/utils/CareViewSummary";

export const CareViewPatientsSummary = (props) => {
  const [setSlotDetails] = useState([]);
  const { patientsSummary } = props;

  const fetchSlots = async (patientUuids, currentTime, endTime) => {
    const response = await getSlotsForPatientsAndTime(
      patientUuids,
      currentTime,
      endTime
    );
    setSlotDetails(response.data);
  };

  useEffect(() => {
    if (patientsSummary.length > 0) {
      const patientUuids = patientsSummary.map(
        (patient) => patient.patientDetails.uuid
      );
      const currentTime = Math.floor(Date.now() / 1000);

      const getTimeInFuture = (offsetHours) => {
        const futureTime = new Date();
        futureTime.setHours(futureTime.getHours() + offsetHours);
        const futureTimeUnix = Math.floor(futureTime.getTime() / 1000);
        return futureTimeUnix.toString();
      };
      const endTime = getTimeInFuture(2);
      fetchSlots(patientUuids, currentTime, endTime);
    }
  }, [patientsSummary]);

  return (
    <div>
      <table className={"care-view-patient-table"}>
        <tbody>
          {patientsSummary.map((patientSummary, idx) => {
            const { patientDetails, bedDetails } = patientSummary;
            const { person } = patientDetails;
            return (
              <tr key={idx} className={"patient-row-container"}>
                <td className={"patient-details-container"}>
                  <div className={"care-view-patient-details"}>
                    <div>
                      <HospitalBed16 />
                      &nbsp;
                      <span>{bedDetails.bedNumber}</span>&nbsp;
                      <Link href={"#"} inline={true}>
                        <span>{patientDetails.display.split(" ")[0]}</span>
                      </Link>
                    </div>
                    <div>
                      <FormattedMessage
                        id={"PATIENT"}
                        defaultMessage={"Patient"}
                      />
                      : <span>{person.display}</span>&nbsp;(
                      <span>{person.gender}</span>)&nbsp;.
                      <span>{person.age}</span>
                      <FormattedMessage
                        id={"AGE_YEARS_LABEL"}
                        defaultMessage={"yrs"}
                      />
                    </div>
                  </div>
                </td>
                <td style={{ minWidth: "650px" }}></td>
                <td style={{ minWidth: "650px" }}></td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
};

CareViewPatientsSummary.propTypes = {
  patientsSummary: propTypes.array,
};
