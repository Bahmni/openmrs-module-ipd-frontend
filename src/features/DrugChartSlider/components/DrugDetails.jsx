import React from "react";
import { TextInput } from "carbon-components-react";
import {
  NumberInputCarbon,
  DropdownCarbon,
  DatePickerCarbon,
} from "bahmni-carbon-ui";
import PropTypes from "prop-types";
import moment from "moment";

export const DrugDetails = ({ hostData }) => {
  return (
    <>
      <TextInput
        id="drug-name"
        className="drug-name"
        type="text"
        value={hostData?.drugOrder?.drugOrder?.drug?.name}
        labelText="Drug Name"
        disabled
      />
      <div className="inline-field">
        <div className="dose-field-with-units">
          <NumberInputCarbon
            id={"Dropdown"}
            onChange={() => {}}
            style={{ width: "50%" }}
            label={"Dose"}
            value={hostData?.drugOrder?.uniformDosingType?.dose}
            isDisabled={true}
          />
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={" "}
            style={{ paddingLeft: "10px" }}
            selectedValue={hostData?.drugOrder?.uniformDosingType?.doseUnits}
            options={[]}
            isDisabled={true}
          />
        </div>
        <div className="route">
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={"Route"}
            selectedValue={hostData?.drugOrder?.route}
            options={[]}
            isDisabled={true}
          />
        </div>
      </div>
      <div className="inline-field">
        <div className="duration-field-with-units">
          <NumberInputCarbon
            id={"Dropdown"}
            onChange={() => {}}
            label={"Duration"}
            value={hostData?.drugOrder?.drugOrder?.duration}
            isDisabled={true}
          />
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={" "}
            selectedValue={hostData?.drugOrder?.drugOrder?.durationUnits}
            options={[]}
            isDisabled={true}
          />
        </div>
        <DatePickerCarbon
          id={"Dropdown"}
          onChange={() => {}}
          titleText={"Start Date"}
          title={"Start Date"}
          value={moment(hostData?.drugOrder?.drugOrder?.scheduledDate).format(
            "MM/DD/YYYY"
          )}
          isDisabled={true}
        />
      </div>
      <div className="frequency">
        <DropdownCarbon
          id={"DropdownFrequency"}
          onChange={() => {}}
          titleText={"Frequency"}
          selectedValue={hostData?.drugOrder?.uniformDosingType?.frequency}
          options={[]}
          isDisabled={true}
        />
      </div>
    </>
  );
};

DrugDetails.propTypes = {
  hostData: PropTypes.shape({
    drugOrder: PropTypes.shape({
      drug: PropTypes.shape({
        name: PropTypes.string,
      }),
      uniformDosingType: PropTypes.shape({
        dose: PropTypes.string,
        doseUnits: PropTypes.string,
        frequency: PropTypes.string,
      }),
      route: PropTypes.string,
      duration: PropTypes.string,
      durationUnit: PropTypes.string,
      scheduledDate: PropTypes.string,
    }),
  }),
};
