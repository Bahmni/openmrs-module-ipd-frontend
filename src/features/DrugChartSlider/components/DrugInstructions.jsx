import React from "react";
import { TextArea } from "carbon-components-react";
import PropTypes from "prop-types";

export const DrugInstructions = ({ hostData }) => {
  return (
    <>
      <div className="instructions">
        <TextArea
          className="instruction"
          readOnly
          type="text"
          rows={1}
          value={hostData?.drugOrder?.instructions}
          labelText="Instruction"
          disabled
        />
      </div>
      <div className="additional-instructions">
        <TextArea
          className="additional-instruction"
          readOnly
          type="text"
          rows={1}
          value={hostData?.drugOrder?.additionalInstructions}
          labelText="Additional Instruction"
          disabled
        />
      </div>
    </>
  );
};

DrugInstructions.propTypes = {
  hostData: PropTypes.shape({
    drugOrder: PropTypes.shape({
      instructions: PropTypes.string,
      additionalInstructions: PropTypes.string,
    }),
  }),
};
