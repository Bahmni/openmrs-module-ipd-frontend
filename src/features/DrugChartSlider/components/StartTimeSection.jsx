// StartTimeSection.js
import React from "react";
import { FormattedMessage } from "react-intl";
import { TimePicker, TimePicker24Hour } from "bahmni-carbon-ui";
import PropTypes from "prop-types";
import {
  invalidTimeText12Hour,
  invalidTimeText24Hour,
} from "../utils/DrugChartSliderUtils";

export const StartTimeSection = ({
  startTime,
  handleStartTime,
  showEmptyStartTimeWarning,
  showStartTimeBeyondNextDoseWarning,
  showStartTimePassedWarning,
  enable24HourTimers,
}) => {
  return (
    <div className="start-time">
      {enable24HourTimers ? (
        <>
          <TimePicker24Hour
            data-modal-primary-focus
            labelText={"Start Time"}
            id={"start-time"}
            onChange={(time) => handleStartTime(time)}
            isRequired={true}
            defaultTime={startTime}
            width="80%"
            invalidText={invalidTimeText24Hour}
          />
          {showStartTimePassedWarning && (
            <p className="time-warning">
              <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_PASSED"></FormattedMessage>
            </p>
          )}
          {showStartTimeBeyondNextDoseWarning && (
            <p className="time-warning">
              <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_BEYOND_NEXT_DOSE"></FormattedMessage>
            </p>
          )}
          {showEmptyStartTimeWarning && (
            <p className="time-error">
              <FormattedMessage id="DRUG_CHART_MODAL_EMPTY_START_TIME_WARNING"></FormattedMessage>
            </p>
          )}
        </>
      ) : (
        <>
          <TimePicker
            id={"start-time"}
            onChange={(time) => handleStartTime(time)}
            defaultTime={startTime}
            labelText={"Start Time"}
            isRequired={true}
            invalidText={invalidTimeText12Hour}
          />
          {showStartTimePassedWarning && (
            <p className="time-warning">
              <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_PASSED"></FormattedMessage>
            </p>
          )}
          {showStartTimeBeyondNextDoseWarning && (
            <p className="time-warning">
              <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_BEYOND_NEXT_DOSE"></FormattedMessage>
            </p>
          )}
          {showEmptyStartTimeWarning && (
            <p className="time-error">
              <FormattedMessage id="DRUG_CHART_MODAL_EMPTY_START_TIME_WARNING"></FormattedMessage>
            </p>
          )}
        </>
      )}
    </div>
  );
};

StartTimeSection.propTypes = {
  startTime: PropTypes.string.isRequired,
  handleStartTime: PropTypes.func.isRequired,
  showEmptyStartTimeWarning: PropTypes.bool.isRequired,
  showStartTimeBeyondNextDoseWarning: PropTypes.bool.isRequired,
  showStartTimePassedWarning: PropTypes.bool.isRequired,
  enable24HourTimers: PropTypes.bool.isRequired,
};
