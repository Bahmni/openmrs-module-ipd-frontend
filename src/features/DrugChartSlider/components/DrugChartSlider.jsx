import React, { useContext, useEffect, useState } from "react";
import { TextArea } from "carbon-components-react";
import moment from "moment";
import PropTypes from "prop-types";
import { FormattedMessage } from "react-intl";
import { I18nProvider } from "../../i18n/I18nProvider";
import SideBarPanel from "../../SideBarPanel/components/SideBarPanel";
import "../styles/DrugChartSlider.scss";
import { medicationFrequency } from "../../../constants";
import { SaveAndCloseButtons } from "../../SaveAndCloseButtons/components/SaveAndCloseButtons";
import { SliderContext } from "../../../context/SliderContext";
import {
  isTimePassed,
  validateSchedules,
  saveMedication,
  updateStartTimeBasedOnFrequency,
  isInvalidTimeTextPresent,
  getUTCTimeEpoch,
} from "../utils/DrugChartSliderUtils";
import { DrugDetails } from "./DrugDetails";
import { DrugInstructions } from "./DrugInstructions";
import { StartTimeSection } from "./StartTimeSection";
import { ScheduleSection } from "./ScheduleSection";

const DrugChartSlider = (props) => {
  const { title, hostData, hostApi, setDrugChartNotes, drugChartNotes } = props;
  const enableSchedule = hostData?.scheduleFrequencies.find(
    (frequency) =>
      frequency.name === hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enableStartTime = hostData?.startTimeFrequencies.includes(
    hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enable24HourTimers = hostData?.enable24HourTimers || false;
  const isAutoFill = Boolean(enableSchedule?.scheduleTiming);

  const { setSliderContentModified } = useContext(SliderContext);
  const updateSliderContentModified = (value) => {
    setSliderContentModified((prev) => {
      return {
        ...prev,
        treatments: value,
      };
    });
  };
  const [firstDaySlotsMissed, setFirstDaySlotsMissed] = useState(0);
  const [startTime, setStartTime] = useState("");

  const [
    showStartTimeBeyondNextDoseWarning,
    setShowStartTimeBeyondNextDoseWarning,
  ] = useState(false);
  const [showStartTimePassedWarning, setShowStartTimePassedWarning] =
    useState(false);
  const [showEmptyStartTimeWarning, setShowEmptyStartTimeWarning] =
    useState(false);

  const [schedules, setSchedules] = useState([]);
  const [firstDaySchedules, setFirstDaySchedules] = useState([]);
  const [finalDaySchedules, setFinalDaySchedules] = useState([]);

  const [showSchedulePassedWarning, setShowSchedulePassedWarning] = useState(
    []
  );
  const [showScheduleOrderWarning, setShowScheduleOrderWarning] =
    useState(false);
  const [showEmptyScheduleWarning, setShowEmptyScheduleWarning] =
    useState(false);

  const [
    showFirstDaySchedulePassedWarning,
    setShowFirstDaySchedulePassedWarning,
  ] = useState([]);
  const [
    showFirstDayScheduleOrderWarning,
    setShowFirstDayScheduleOrderWarning,
  ] = useState(false);
  const [
    showEmptyFirstDayScheduleWarning,
    setShowEmptyFirstDayScheduleWarning,
  ] = useState(false);

  const [
    showFinalDayScheduleOrderWarning,
    setShowFinalDayScheduleOrderWarning,
  ] = useState(false);
  const [
    showEmptyFinalDayScheduleWarning,
    setShowEmptyFinalDayScheduleWarning,
  ] = useState(false);

  const handleFirstDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...firstDaySchedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setFirstDaySchedules(newScheduleArray);
    if (!isInvalidTimeTextPresent(enable24HourTimers)) {
      setShowFirstDaySchedulePassedWarning((prevScheduleWarnings) => {
        const newSchedulePassedWarnings = [...prevScheduleWarnings];
        newSchedulePassedWarnings[index] = isTimePassed(newSchedule);
        return newSchedulePassedWarnings;
      });
    }
  };

  const handleSubsequentDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...schedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setSchedules(newScheduleArray);
    if (!isInvalidTimeTextPresent(enable24HourTimers)) {
      setShowSchedulePassedWarning((prevScheduleWarnings) => {
        const newSchedulePassedWarnings = [...prevScheduleWarnings];
        newSchedulePassedWarnings[index] = isTimePassed(newSchedule);
        return newSchedulePassedWarnings;
      });
    }
  };

  const handleFinalDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...finalDaySchedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setFinalDaySchedules(newScheduleArray);
  };

  const handleScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(schedules);
    setShowEmptyScheduleWarning(!isValid && warningType === "empty");
    setShowScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidSchedule = async () => {
    const { isValid, warningType } = await handleScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const handleFirstDayScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(
      firstDaySchedules.filter(
        (firstDaySchedule) => firstDaySchedule != "hh:mm"
      )
    );
    setShowEmptyFirstDayScheduleWarning(!isValid && warningType === "empty");
    setShowFirstDayScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidFirstDaySchedule = async () => {
    const { isValid, warningType } = await handleFirstDayScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const handleFinalDayScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(finalDaySchedules);
    setShowEmptyFinalDayScheduleWarning(!isValid && warningType === "empty");
    setShowFinalDayScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidFinalDaySchedule = async () => {
    const { isValid, warningType } = await handleFinalDayScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const isStartTimeExceedingFrequency = (time, frequency) => {
    const currentTime = moment();
    const enteredTime = enable24HourTimers
      ? moment(time, "HH:mm")
      : moment(time, "hh:mm A");
    const updatedTime = updateStartTimeBasedOnFrequency(frequency, currentTime);
    return enteredTime.isAfter(updatedTime);
  };

  const handleStartTime = async (time) => {
    updateSliderContentModified(true);
    if (time === "") {
      setShowEmptyStartTimeWarning(true);
      setShowStartTimeBeyondNextDoseWarning(false);
      setShowStartTimePassedWarning(false);
      return;
    } else {
      setShowEmptyStartTimeWarning(false);
    }
    if (
      (enable24HourTimers && !moment(time, "HH:mm", true).isValid()) ||
      (!enable24HourTimers && !moment(time, "hh:mm A", true).isValid())
    ) {
      setStartTime(time);
      return;
    }
    isStartTimeExceedingFrequency(
      time,
      hostData?.drugOrder?.uniformDosingType?.frequency
    )
      ? setShowStartTimeBeyondNextDoseWarning(true)
      : setShowStartTimeBeyondNextDoseWarning(false);
    isTimePassed(time)
      ? setShowStartTimePassedWarning(true)
      : setShowStartTimePassedWarning(false);

    enable24HourTimers
      ? setStartTime(time)
      : setStartTime(moment(time, "hh:mm A"));
  };

  const createDrugChartPayload = () => {
    var payload = {
      providerUuid: hostData?.drugOrder?.provider?.uuid,
      patientUuid: hostData?.patientId,
      orderUuid: hostData?.drugOrder?.uuid,
      slotStartTime: null,
      firstDaySlotsStartTime: null,
      dayWiseSlotsStartTime: null,
      remainingDaySlotsStartTime: null,
      comments: drugChartNotes,
      medicationFrequency: "",
    };
    if (enableStartTime) {
      const startTimeUTCEpoch = getUTCTimeEpoch(
        startTime,
        enable24HourTimers,
        hostData?.drugOrder?.scheduledDate
      );
      payload.slotStartTime = startTimeUTCEpoch;
      payload.medicationFrequency =
        medicationFrequency.START_TIME_DURATION_FREQUENCY;
    }
    if (enableSchedule) {
      const nextScheduleDate = 24 * 60 * 60;
      const finalScheduleDate =
        nextScheduleDate * hostData?.drugOrder?.duration;

      const firstDaySchedulesUTCTimeEpoch = firstDaySchedules.reduce(
        (result, schedule) => {
          if (schedule !== "hh:mm") {
            result.push(
              getUTCTimeEpoch(
                schedule,
                enable24HourTimers,
                hostData?.drugOrder?.scheduledDate
              )
            );
          }
          return result;
        },
        []
      );

      const schedulesUTCTimeEpoch = schedules.map((schedule) =>
        getUTCTimeEpoch(
          schedule,
          enable24HourTimers,
          hostData?.drugOrder?.scheduledDate
        )
      );

      const finalDaySchedulesUTCTimeEpoch = finalDaySchedules.map((schedule) =>
        getUTCTimeEpoch(
          schedule,
          enable24HourTimers,
          hostData?.drugOrder?.scheduledDate
        )
      );

      payload.firstDaySlotsStartTime =
        firstDaySlotsMissed > 0 ? firstDaySchedulesUTCTimeEpoch : [];
      payload.dayWiseSlotsStartTime = firstDaySchedules.some(
        (schedule) => schedule == "hh:mm"
      )
        ? schedulesUTCTimeEpoch.map((schedules) => schedules + nextScheduleDate)
        : schedulesUTCTimeEpoch;
      const remainingDaySlotsStartTime = finalDaySchedulesUTCTimeEpoch.map(
        (schedules) => schedules + finalScheduleDate
      );

      const remainingDaySlotsTime = remainingDaySlotsStartTime.slice(
        0,
        firstDaySlotsMissed
      );
      payload.remainingDaySlotsStartTime = remainingDaySlotsTime;
      payload.medicationFrequency =
        medicationFrequency.FIXED_SCHEDULE_FREQUENCY;
    }
    return payload;
  };

  const validateSave = async () => {
    if (isInvalidTimeTextPresent(enable24HourTimers)) return false;
    if (enableSchedule) {
      const validFirstDaySchedules = await isValidFirstDaySchedule();
      const validSchedules = await isValidSchedule();
      const validFinalDaySchedules = await isValidFinalDaySchedule();
      if (!validFirstDaySchedules || !validSchedules || !validFinalDaySchedules)
        return false;
      return true;
    }
    if (enableStartTime) {
      if (!startTime) {
        setShowEmptyStartTimeWarning(true);
        return false;
      }
      return true;
    }
  };

  useEffect(() => {
    if (isAutoFill) {
      const scheduleTimings = enable24HourTimers
        ? enableSchedule?.scheduleTiming
        : enableSchedule?.scheduleTiming.map((time) => moment(time, "hh:mm A"));
      let finalScheduleCount = 0;
      scheduleTimings.forEach((schedule) => {
        if (isTimePassed(schedule)) {
          setFirstDaySchedules((prevSchedules) => [...prevSchedules, "hh:mm"]);
          finalScheduleCount = finalScheduleCount + 1;
          setFirstDaySlotsMissed((prevSlotNumber) => prevSlotNumber + 1);
        } else {
          setFirstDaySchedules((prevSchedules) => [...prevSchedules, schedule]);
        }
      });
      setSchedules(scheduleTimings || []);
      finalScheduleCount > 0 &&
        setFinalDaySchedules(scheduleTimings.slice(firstDaySlotsMissed) || []);
      if (finalScheduleCount == enableSchedule?.frequencyPerDay) {
        setFirstDaySchedules([]);
        setFinalDaySchedules([]);
        setFirstDaySlotsMissed(0);
      }
    } else {
      const defaultSchedules = Array.from(
        { length: enableSchedule?.frequencyPerDay },
        () => ""
      );
      setSchedules(defaultSchedules);
    }
  }, [isAutoFill, enable24HourTimers, enableSchedule]);

  const handleSave = async () => {
    const performSave = await validateSave();
    if (performSave) {
      const medication = createDrugChartPayload();
      const response = await saveMedication(medication);
      response.status === 200 ? hostApi.onModalSave?.() : null;
    }
  };

  const handleCancel = () => {
    hostApi.onModalCancel?.();
  };

  const handleClose = () => {
    hostApi.onModalClose?.();
  };

  const handleNotes = (e) => {
    updateSliderContentModified(true);
    setDrugChartNotes(e.target.value);
  };

  return (
    <I18nProvider>
      <SideBarPanel
        title={
          title ? title : <FormattedMessage id="DRUG_CHART_MODAL_HEADER" />
        }
        closeSideBar={handleClose}
      >
        <div style={{ padding: "20px", paddingBottom: "120px" }}>
          <DrugDetails hostData={hostData} />
          <ScheduleSection
            enableSchedule={enableSchedule}
            firstDaySlotsMissed={firstDaySlotsMissed}
            firstDaySchedules={firstDaySchedules}
            schedules={schedules}
            finalDaySchedules={finalDaySchedules}
            handleFirstDaySchedule={handleFirstDaySchedule}
            handleSubsequentDaySchedule={handleSubsequentDaySchedule}
            handleFinalDaySchedule={handleFinalDaySchedule}
            showFirstDayScheduleOrderWarning={showFirstDayScheduleOrderWarning}
            showEmptyFirstDayScheduleWarning={showEmptyFirstDayScheduleWarning}
            showFirstDaySchedulePassedWarning={
              showFirstDaySchedulePassedWarning
            }
            showScheduleOrderWarning={showScheduleOrderWarning}
            showEmptyScheduleWarning={showEmptyScheduleWarning}
            showFinalDayScheduleOrderWarning={showFinalDayScheduleOrderWarning}
            showEmptyFinalDayScheduleWarning={showEmptyFinalDayScheduleWarning}
            showSchedulePassedWarning={showSchedulePassedWarning}
            enable24HourTimers={enable24HourTimers}
          />
          {enableStartTime && (
            <StartTimeSection
              startTime={startTime}
              handleStartTime={handleStartTime}
              showEmptyStartTimeWarning={showEmptyStartTimeWarning}
              showStartTimeBeyondNextDoseWarning={
                showStartTimeBeyondNextDoseWarning
              }
              showStartTimePassedWarning={showStartTimePassedWarning}
              enable24HourTimers={enable24HourTimers}
            />
          )}
          <DrugInstructions hostData={hostData} />
          <div className="notes-sections">
            <TextArea
              data-testid="notes-section"
              className="notes-section"
              type="text"
              rows={3}
              value={drugChartNotes}
              onChange={(e) => handleNotes(e)}
              labelText="Notes"
            />
          </div>
        </div>
        <SaveAndCloseButtons
          onSave={() => handleSave()}
          onClose={() => handleCancel()}
        />
      </SideBarPanel>
    </I18nProvider>
  );
};

DrugChartSlider.propTypes = {
  title: PropTypes.string.isRequired,
  hostData: PropTypes.object.isRequired,
  hostApi: PropTypes.object.isRequired,
  setDrugChartNotes: PropTypes.func.isRequired,
  drugChartNotes: PropTypes.string.isRequired,
};

export default DrugChartSlider;
