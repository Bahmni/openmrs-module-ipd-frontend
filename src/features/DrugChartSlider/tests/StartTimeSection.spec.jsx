import React from "react";
import { render } from "@testing-library/react";
import { StartTimeSection } from "../components/StartTimeSection";

const testProps = {
  startTime: "11:00",
  handleStartTime: jest.fn(),
  showEmptyStartTimeWarning: false,
  showStartTimeBeyondNextDoseWarning: false,
  showStartTimePassedWarning: false,
  enable24HourTimers: false,
};

describe("StartTimeSection", () => {
  it("should match snapshot", () => {
    const { container } = render(<StartTimeSection props={testProps} />);
    expect(container).toMatchSnapshot();
  });
});
