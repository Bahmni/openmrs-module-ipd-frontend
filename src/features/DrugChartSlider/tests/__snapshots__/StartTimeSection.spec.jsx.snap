// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartTimeSection should match snapshot 1`] = `
<div>
  <div
    class="start-time"
  >
    <div
      class="bx--form-item"
    >
      <label
        class="bx--label"
        for="time-selector"
      >
        <div
          class="titleText"
        >
          <span>
            Start Time 
          </span>
          <span
            class="required"
          >
            *
          </span>
        </div>
      </label>
      <div
        class="bx--time-picker"
      >
        <div
          class="bx--time-picker__input"
        >
          <input
            autocomplete="off"
            class="bx--time-picker__input-field bx--text-input"
            id="time-selector"
            maxlength="5"
            pattern="(1[012]|[1-9]):[0-5][0-9](\\s)?"
            placeholder="hh:mm"
            style="width: 72px; padding: 0px 0px 0px 1rem;"
            type="text"
            value=""
          />
        </div>
        <div
          class="bx--select bx--time-picker__select"
        >
          <label
            class="bx--label bx--visually-hidden"
            for="time-picker-select-1"
          >
            Choose a time
          </label>
          <select
            aria-label="open list of options"
            class="bx--select-input"
            id="time-picker-select-1"
          >
            <option
              class="bx--select-option"
              value="AM"
            >
              AM
            </option>
            <option
              class="bx--select-option"
              value="PM"
            >
              PM
            </option>
          </select>
          <svg
            aria-label="open list of options"
            class="bx--select__arrow"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
